File: i-pro-b/l1/prob/f1.py
n: int
m: int

graph: list
used: list
match: list


def kuhn(v):

    global graph, used, match

    if (used[v]):
        return False;

    used[v] = True;

    for vt in graph[v]:
        if match[vt] == -1 or kuhn(match[vt]):
            match[vt] = v
            return True
    
    return False


def main():

    global n, m
    global graph, used, match

    n, m = map(int, input().split())

    graph = [list() for _ in range(n)]
    match = [-1 for _ in range(n)]

    for _ in range(m):
        x, y = map(int, input().split())
        graph[x - 1].append(y - 1)

    for v in range(n):
        used = [False for _ in range(n)]
        kuhn(v)

    print(n - len(match) + match.count(-1))


if __name__ == '__main__':
    main()

File: i-pro-b/l1/prob/f1.cxx
# include <iostream>
# include <vector>


bool kuhn (size_t v, 
    std::vector<std::vector<int>>& graph, 
    std::vector<bool>& used,
    std::vector<int>& match) {
    /*
    if (used[v]) return false;
    used[v] = true;

    for (auto vt : graph[v]) {
        if (match[vt] == -1 || kuhn (match[vt], graph, used, match)) {
            match[vt] = v;
            return true;
        }
    }
    */

    return false;
}


int iprob_f1_main () {

    size_t n, m;
    std::cin >> n >> m;

    auto graph = std::vector<std::vector<int>> (n, std::vector<int>{});
    auto match = std::vector<int> (n, -1);

    int x, y;
    for (size_t i = 0; i < m; ++i) {
        std::cin >> x >> y;
        graph[x - 1].push_back (y - 1);
    }

    for (size_t v = 0; v < n; ++v) {
        auto used = std::vector<bool> (n, false);
        kuhn (v, graph, used, match);
    }

    int l = 0;
    for (auto i : match) if (i == -1) ++l;

    std::cout << n - match.size () + l;

    return 0;
}